sourcegraph/cody
# VScode: Enhanced context settings dialog and add local embeddings

Adds a popup in the chat box which lets you control enhanced context. You can start local embeddings indexing from there, if there's no remote embeddings index.

Note, this needs a cody-engine version roll before landing.

Co-authored with @toolmantim 

## Caveats

- Local embeddings is only enabled when `"cody.experimental.simpleChatContext": true` but it can work with classic chat context, see comments around ContextProvider.init.
- Simple chat context tries to not add enhanced context for subsequent messages. I have hackily made that work and put some TODOs for @beyang to look at.
- Apparently races at startup? Sometimes the enhanced context dialog displays no context sources.
- Assumes that the workspace root is the repository root.
- Error reporting from embeddings generation failure is limited, and jobs don't restart.
- Various caveats for local embeddings in cody-engine are tracked in a separate issue tracker.

## Test plan

```
pnpm test:unit
```

Manual integration test:

1. Build cody-engine after f834300. See elsewhere for details.
2. Set the following user settings, then run the extension:

```
  "cody.debug.enable": true,
  "cody.debug.verbose": true,
  "cody.experimental.cody-engine.path": "path/to/cody-engine/binary",
  "cody.experimental.chatPanel": true,
  "cody.experimental.simpleChatContext": true,
```

3. Open a folder that was indexed by Cody App (specifically, that has a git remote with a default fetch URL that matches something indexed by App on your computer.) Start a new chat. Click on the enhanced context selector, Embeddings should have a checkmark. Make a query, embeddings results should appear.
4. Open a folder containing a git repo that is indexed on dotcom. Start a new chat. Click on the enhanced context selector, Embeddings should have a checkmark with a note that they are "inherited" etc.
5. Open a folder that does not contain a git repo. Start a new chat. No sources should appear in the enhanced context selector.
6. Open a folder that contains a git repo that was not indexed by App, nor is indexed on dotcom. Start a new chat. Open the enhanced context selector. A button to enable embeddings should appear. Click the button. The status bar should show embeddings progress. If the job completes, make a query, embeddings results should appear. If the job fails, reload the window and chat/query and you should get partial results from the index.

To reset local embeddings indexes, if that is desired, delete `~/Library/Caches/com.sourcegraph.cody/embeddings`.
